dspy/teleprompt/bootstrap_finetune.py

```diff
@@ -0,0 +1,261 @@
+            if training_key not in key_to_data:
+                train_data, data_format = self._prepare_finetune_data(trace_data=trace_data, lm=pred.lm, pred_ind=data_pred_ind)
+                finetune_kwargs = dict(lm=pred.lm, train_data=train_data, train_kwargs=self.train_kwargs[pred.lm], data_format=data_format)
+                key_to_data[training_key] = finetune_kwargs
+        logger.info("Done.")
+        
+        logger.info("Starting LM fine-tuning...")
+        # TODO(feature): We could run batches of fine-tuning jobs in sequence
+        # to avoid exceeding the number of threads.
+        err = f"BootstrapFinetune requires `num_threads` to be bigger than or equal to the number of fine-tuning jobs. There are {len(key_to_data)} fine-tuning jobs to start, but the number of threads is: {self.num_threads}! If the `multitask` flag is set to False, the number of fine-tuning jobs will be equal to the number of predictors in the student program. If the `multitask` flag is set to True, the number of fine-tuning jobs will be equal to: 1 if there is only a context LM, or the number of unique LMs attached to the predictors in the student program. In any case, the number of fine-tuning jobs will be less than or equal to the number of predictors."
+        assert len(key_to_data) <= self.num_threads, err
+        logger.info(f"{len(key_to_data)} fine-tuning jobs to start.")
+        key_to_lm = self.finetune_lms(key_to_data)
+        logger.info("Done.")
+
+        logger.info("Updating the student program with the fine-tuned LMs...")
+        for pred_ind, pred in enumerate(student.predictors()):
+            data_pred_ind = None if self.multitask else pred_ind
+            training_key = (pred.lm, data_pred_ind)
+            pred.lm = key_to_lm[training_key]
+        logger.info("Done.")
+
+        return student
+
+    @staticmethod
+    def finetune_lms(finetune_dict) -> Dict[Any, LM]:
+        num_jobs = len(finetune_dict)
+        logger.info(f"Starting {num_jobs} fine-tuning jobs...")
+        # TODO(nit) Pass an identifier to the job so that we can tell the logs
+        # coming from different fine-tune threads.
+
+        key_to_job = {}
+        for key, finetune_kwargs in finetune_dict.items():
+            lm = finetune_kwargs.pop("lm")
+            key_to_job[key] = lm.finetune(**finetune_kwargs)
+        
+        key_to_lm = {}
+        for ind, (key, job) in enumerate(key_to_job.items()):
+            key_to_lm[key] = job.result()
+            job.thread.join()
+            logger.info(f"Job {ind + 1}/{num_jobs} completed.")
+
+        return key_to_lm
+
+    def _prepare_finetune_data(self, trace_data: List[Dict[str, Any]], lm: LM, pred_ind: Optional[int] = None):
+        # TODO(nit) Log dataset details/size
+        if self.metric:
+            trace_data = [d for d in trace_data if d["score"]]
+
+        data = []
+        adapter = self.adapter[lm] or lm.infer_adapter()
+        data_format = infer_data_format(adapter)
```

Comment:
Instead of having constructor take in adapters, and infer the data format from the adapter. Shall we set `data_format` in the constructor?

Reply:
I think the interaction here could be improved, but I don't have concrete proposals.

Either way, we would need to access an "adapter", that may not be directly inferred from the `data_format` (Here it can be inferred, but this is not true for all possible data_formats we could have, say `prompt`, `chosen`, `rejected`) Making a note of this for a future discussion.